plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/DevWatchContext.java

```diff
@@ -0,0 +1,522 @@
+        Path p = watchedDirectories.get(key);
+        if (p == null) {
+            ctx.debug("No more watching key, ignoring change done to " + fileName);
+            return null;
+        } else {
+            Path resolved = p.resolve(fileName);
+            // Fully ignore target dir
+            if (projectBuildDir.equals(resolved)) {
+                return null;
+            }
+            return resolved;
+        }
+
+    }
+
+    private Path getResourcesDir(Path p) {
+        for (Path path : resourceDirectories) {
+            if (p.startsWith(path)) {
+                return path;
+            }
+        }
+        if (p.startsWith(javaDir) && !isJavaFile(p)) {
+            return javaDir;
+        }
+        return null;
+    }
+
+    private boolean isBootableSpecificFile(Path p) {
+        if (cliFiles.contains(p)) {
+            return true;
+        }
+        for (Path path : extraDirectories) {
+            if (p.startsWith(path)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void registerDir(Path dir) throws IOException {
+        registerDir(dir, null);
+    }
+
+    private void registerDir(Path dir, Set<Path> set) throws IOException {
+        Files.walkFileTree(dir, new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                if (!projectBuildDir.equals(dir)) {
+                    watchedDirectories.put(dir.register(watcher, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY), dir);
+                    ctx.debug("[WATCH] watching " + dir);
+                    if (set != null) {
+                        set.add(dir);
+                    }
+                    return FileVisitResult.CONTINUE;
+                } else {
+                    return FileVisitResult.SKIP_SUBTREE;
+                }
+            }
+        });
+    }
+
+    void cleanup() {
+        for (WatchKey k : watchedDirectories.keySet()) {
+            k.cancel();
+        }
+        if (Files.exists(getTargetDirectory())) {
+            IoUtils.recursiveDelete(getTargetDirectory());
+        }
+    }
+
+    BootableAppEventHandler newEventHandler() {
+        BootableAppEventHandler handler;
+        if (isWebApp) {
+            handler = new WebAppEventHandler();
+        } else {
+            if (isJarApp) {
+                handler = new JavaAppEventHandler();
+            } else {
+                throw new RuntimeException("Not supported packaging");
+            }
+        }
+        return handler;
+    }
+
+    void build(boolean autoCompile) throws IOException, MojoExecutionException {
+        rebuild(autoCompile, true, true, true, true);
+    }
+
+    private void rebuild(boolean autoCompile, boolean compile, boolean repackage, boolean redeploy, boolean cleanup) throws IOException, MojoExecutionException {
+        if (cleanup) {
+            ctx.cleanup();
```

Comment:
Just kind of a general comment. We should probably only do the clean up when the process has stopped. Looking at the implementation it deletes paths. If the process is still running these will likely not be deleted on Windows.

Reply:
The cleanup is fired in 1 case. A user deleted a java file, we need to delete the classes in the target/classes and in the deployment and repackage/redeploy. It seems to work on windows, at least all the tests are passing very often ;-)

Reply:
Okay cool. I just remember seeing, maybe not even part of this PR :), a bunch of logs indicating the files couldn't be deleted. I should have checked before I just assumed anything :)